karabo_data/geometry2.py

```diff
@@ -631,6 +695,229 @@ def plot_data(self, modules_data):
 {name}/coffset = {coffset}
 """
 
+class LPD_1MGeometry(DetectorGeometryBase):
+    """Detector layout for LPD-1M
+
+    The coordinates used in this class are 3D (x, y, z), and represent multiples
+    of the pixel size.
+    """
+    pixel_size = 5e-4  # 5e-4 metres == 0.5 mm
+    frag_ss_pixels = 32
+    frag_fs_pixels = 128
+    expected_data_shape = (16, 256, 256)
+
+    @classmethod
+    def from_quad_positions(cls, quad_pos, *, unit=1e-3, asic_gap=None,
+                            panel_gap=None):
+        """Generate an LPD-1M geometry from quadrant positions.
+
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The quadrant positions refer to the corner of each quadrant
+        where module 4, tile 16 is positioned.
+        This is the corner of the last pixel as the data is stored.
+        In the initial detector layout, the corner positions are for the top
+        right corner of the quadrant, looking into the beam.
+
+        Parameters
+        ----------
+        quad_pos: list of 2-tuples
+          (x, y) coordinates of the last corner (the one by module 4) of each
+          quadrant.
+        unit: float, optional
+          The conversion factor to put the coordinates into metres.
+          The default 1e-3 means the numbers are in millimetres.
+        asic_gap: float, optional
+          The gap between adjacent tiles/ASICs. The default is 4 pixels.
+        panel_gap: float, optional
+          The gap between adjacent modules/panels. The default is 4 pixels.
+        """
+        px_conversion = unit / cls.pixel_size
+        asic_gap_px = 4 if (asic_gap is None) else asic_gap * px_conversion
+        panel_gap_px = 4 if (panel_gap is None) else panel_gap * px_conversion
+
+        tile_size = np.array([cls.frag_fs_pixels, cls.frag_ss_pixels, 0])
+
+        panels_across = [-1, -1, 0, 0]
+        panels_up = [0, -1, -1, 0]
+        modules = []
+        for p in range(16):
+            quad = p // 4
+            quad_corner_x = quad_pos[quad][0] * px_conversion
```

Comment:
This quad_position is wrt to centre? Is the hole in detector the centre?

Reply:
Me and Martin were trying to understand the quad positions we use as of now

- quad_pos = [(-11.4, -299), (11.5, -8), 
                (-254.5, 16), (-278.5, -275)]

In the docstring you wrote "In the initial detector layout, the corner positions are for the top right corner of the quadrant, looking into the beam."  Somehow this doesn't look consistent if the centre of detector is the origin and y axis points up, x axis points left and z axis point into the detector center.

Reply:
Yup, I should make it clear that the origin is in the centre of the detector. I think the idea is that it's the point where the beam intersects with the detector plane, but I imagine you have to do some refinement if you want a precise beam centre.

The quadrant positions we're using at the moment are backwards (x axis beam-right, y axis down), like the coordinates in the HDF5 file. The new machinery takes coordinates in the standard directions (x axis beam-left, y axis up), so all the numbers have to be inverted. I've done that in the examples:

https://github.com/European-XFEL/karabo_data/pull/107/files#diff-90dbb13e16d6833f559126522a31bc42R46

Reply:
Are you sure that quad_pos in the notebook are still correct? We were trying to reconstruct the quad positions based on these coordinates and it does not add up.

Reply:
looking at geom.inspect shows that the quadrants starting form the top left and going counter clockwise to the top right, but according to the detector intro https://in.xfel.eu/readthedocs/docs/detector-documentation/en/latest/lpd/source/introduction.html the quadrants should start from the top right going clockwise to the to left

Reply:
Dude, when will you be here? I guess Ebad an me are too stupid. Maybe you can show us once your back in HH?

Reply:
The diagram in the detector intro is shown in 'beam direction', i.e. as if you were standing with the beam going through you, looking at the detector. `geometry.inspect()` plots it the other way, as if you were looking into the beam (this way follows the usual convention of plotting x increasing left to right). So left/right and clockwise/anticlockwise are flipped.

This makes my head hurt too. :wink: 

For the `plot_data_fast` function, I added a `frontview=True` parameter to flip the image to match the orientation in the diagram ([details above](https://github.com/European-XFEL/karabo_data/pull/107#issuecomment-462318003)). I was going to do the same for `inspect()` in a separate PR to keep the changes here smaller, but if you'd find it easier to review with that, I can add it to this PR.

Reply:
I do not understand the benefit of adding a "center" for the detector since it differs the top-left-corner-origin convention of displaying an image. It increases the burden of our brains since we may need to digest the images with different origins at the same time. 

In reality, the beam usually does not pass through exactly the detector center. Therefore, having a beam position of (2, 3) is not more convenient than having a beam position of (602, 603). In addition, in some experiments, the beam does not even pass through the hole (it bypasses the detector). 

Moreover, speaking of pyFAI, it adopts the bottom-left-corner-origin convention. It means that we may be facing three different origins at the same time :(

Reply:
The CrystFEL geometry already seems to use the centre as the origin, as do the quadrant positions we got to use with the XFEL geometry files. So it's not a new convention we're adding.

An 'image style' origin in the top left doesn't make sense given XFEL's standard coordinate system, because we have y increasing upwards and x increasing beam-left.

Reply:
Nevertheless, the Karabo scenes always display images with the origin being the top left corner. XFEL / CrystFEL's conventions are both implementation details which users may not care. The displayed image is the interface that affects the users. Of course, one could argue CrystFEL is the user. Generally, for people who is doing data analysis, they received the image as an array, and the array index starts from 0.

Reply:
Displaying images is only one use case: you might also save the geometry to a CrystFEL file, or convert it to a PyFAI distortion array. Or an XFEL file, once that's implemented. Shifting the origin to the topmost/leftmost pixels throws away information about where the centre is, which we need in order to write out geometry files comparable to those we read.

Also, if you define the origin in line with the topmost/leftmost pixels, then moving the topmost or leftmost panels in the geometry means moving all the other panels, because the origin shifts. Obviously that's possible if there's no other choice, but it's not elegant, and it makes comparing different geometries harder.

Reply:
Thanks for the explanation! In principle, displaying images can be totally decoupled from saving geometry, converting to something else or any other internal manipulations. But if it is an expert interface, then it makes sense to also display the implemented image coordinate system.